Gmail and WebKit both made writing suggestions!

lilyspiniolas opened the pull request:
#### 5fff6bf0921d978efc01e51fd8137351ad3948e0
<pre>
Gmail and WebKit both made writing suggestions!
<a href="https://bugs.webkit.org/show_bug.cgi?id=275917">https://bugs.webkit.org/show_bug.cgi?id=275917</a>
<a href="https://rdar.apple.com/128360054">rdar://128360054</a>

Reviewed by Megan Gardner, Wenson Hsieh, Abrar Rahman Protyasha and Richard Robinson.

Added a Quirk for mail.google.com to disable writing suggestions provided from WebKit. This prevents duplicate suggestions appearing since mail.google.com now has their own implementation of suggestions.

* Source/WebCore/dom/Element.cpp:
(WebCore::Element::isWritingSuggestionsEnabled const):
* Source/WebCore/page/Quirks.cpp:
(WebCore::Quirks::shouldDisableWritingSuggestionsByDefaultQuirk const):
* Source/WebCore/page/Quirks.h:

Canonical link: <a href="https://commits.webkit.org/280464@main">https://commits.webkit.org/280464@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/0d0f6d87804037f405b74be05b723516f4c1268c

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/56509 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/35835 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/8981 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/60116 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/6951 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/58635 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/43457 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/7141 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/45787 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/4862 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/58538 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/33709 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/48767 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/26642 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/30472 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/6105 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/5949 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/52452 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/6365 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/61800 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/417 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/6489 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/53045 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/418 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/48829 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/52955 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/12547 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/372 "Passed tests") | | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/31663 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/32748 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/33838 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/32496 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

whsieh commented on Source/WebCore/page/Quirks.cpp:
```diff
@@ -365,6 +365,15 @@ bool Quirks::shouldAvoidUsingIOS13ForGmail() const
 #endif
 }
 
+// mail.google.com rdar://128360054
```
Nit - maybe instead of listing the bug number for the quirk here, we could turn it into a FIXME; something like:

```
// FIXME (<rdar or bugzilla bug>): Remove this quirk once Gmail adopts the `writingsuggestions` attribute.
```

pxlcoder replied:
This is done throughout the file to easily indicate the (possibly multiple sites) reasons behind a quirk.

But a FIXME underneath that is good!

lilyspiniolas replied:
Added the FIXME underneath in https://github.com/WebKit/WebKit/commit/0d0f6d87804037f405b74be05b723516f4c1268c

aprotyas commented on Source/WebCore/page/Quirks.h:
```diff
@@ -98,6 +98,7 @@ class Quirks {
     WEBCORE_EXPORT bool needsYouTubeMouseOutQuirk() const;
 
     WEBCORE_EXPORT bool shouldAvoidUsingIOS13ForGmail() const;
+    WEBCORE_EXPORT bool shouldDisableWritingSuggestionsByDefaultQuirk() const;
```
Nit: this is a method on the `Quirks` class anyway, so a `Quirk` suffix here seems redundant.

lilyspiniolas replied:
Fixed in https://github.com/WebKit/WebKit/commit/0d0f6d87804037f405b74be05b723516f4c1268c

aproskuryakov commented on Source/WebCore/page/Quirks.cpp:
```diff
@@ -365,6 +365,16 @@ bool Quirks::shouldAvoidUsingIOS13ForGmail() const
+    if (!needsQuirks())
+        return false;
+    auto& url = m_document->topDocument().url();
+    return url.host() == "mail.google.com"_s;
```
Is Gmail always at this domain? I was wondering if there was some variation where corporations used their own domain.

pxlcoder replied:
Not certain... this does match the other Gmail quirks in this file though.

lilyspiniolas committed:
Gmail and WebKit both made writing suggestions!
https://bugs.webkit.org/show_bug.cgi?id=275917
rdar://128360054

Reviewed by Megan Gardner, Wenson Hsieh, Abrar Rahman Protyasha and Richard Robinson.

Added a Quirk for mail.google.com to disable writing suggestions provided from WebKit. This prevents duplicate suggestions appearing since mail.google.com now has their own implementation of suggestions.

* Source/WebCore/dom/Element.cpp:
(WebCore::Element::isWritingSuggestionsEnabled const):
* Source/WebCore/page/Quirks.cpp:
(WebCore::Quirks::shouldDisableWritingSuggestionsByDefaultQuirk const):
* Source/WebCore/page/Quirks.h:

Canonical link: https://commits.webkit.org/280464@main
